spark-operator/src/main/java/org/apache/spark/k8s/operator/reconciler/SparkClusterReconciler.java

```diff
@@ -0,0 +1,181 @@
+      SparkCluster sparkCluster, Context<SparkCluster> context, Exception e) {
+    LoggingUtils.TrackedMDC trackedMDC = new LoggingUtils.TrackedMDC();
+    try {
+      context
+          .getRetryInfo()
+          .ifPresent(
+              retryInfo -> {
+                if (log.isErrorEnabled()) {
+                  log.error(
+                      "Failed attempt: {}, last attempt: {}",
+                      retryInfo.getAttemptCount(),
+                      retryInfo.isLastAttempt());
+                }
+              });
+      return ErrorStatusUpdateControl.noStatusUpdate();
+    } finally {
+      trackedMDC.reset();
+    }
+  }
+
+  @Override
+  public Map<String, EventSource> prepareEventSources(EventSourceContext<SparkCluster> context) {
+    EventSource podEventSource =
+        new InformerEventSource<>(
+            InformerConfiguration.from(Pod.class, context)
+                .withSecondaryToPrimaryMapper(
+                    Mappers.fromLabel(Constants.LABEL_SPARK_APPLICATION_NAME))
+                .withLabelSelector(commonResourceLabelsStr())
+                .build(),
+            context);
+    return EventSourceInitializer.nameEventSources(podEventSource);
+  }
+
+  protected List<ClusterReconcileStep> getReconcileSteps(final SparkCluster cluster) {
+    List<ClusterReconcileStep> steps = new ArrayList<>();
+    steps.add(new ClusterValidateStep());
+    steps.add(new ClusterTerminatedStep());
+    switch (cluster.getStatus().getCurrentState().getCurrentStateSummary()) {
+      case Submitted -> steps.add(new ClusterInitStep());
+      case RunningHealthy -> {
+        // There is nothing to do because Spark Cluster is supposed to run infinitely.
+      }
+      default -> steps.add(new ClusterUnknownStateStep());
+    }
+    return steps;
+  }
+
+  /**
+   * Best-effort graceful termination upon delete.
+   *
+   * @param sparkCluster the resource that is marked for deletion
+   * @param context the context with which the operation is executed
+   * @return DeleteControl, with requeue if needed
+   */
+  @Override
+  public DeleteControl cleanup(SparkCluster sparkCluster, Context<SparkCluster> context) {
+    LoggingUtils.TrackedMDC trackedMDC = new LoggingUtils.TrackedMDC();
+    try {
+      log.info("Cleaning up resources for SparkCluster.");
+      SparkClusterContext ctx = new SparkClusterContext(sparkCluster, context, submissionWorker);
+      List<ClusterReconcileStep> cleanupSteps = new ArrayList<>();
+      cleanupSteps.add(new ClusterValidateStep());
+      cleanupSteps.add(new ClusterTerminatedStep());
+      for (ClusterReconcileStep step : cleanupSteps) {
+        ReconcileProgress progress = step.reconcile(ctx, sparkClusterStatusRecorder);
+        if (progress.isCompleted()) {
+          if (progress.isRequeue()) {
+            return DeleteControl.noFinalizerRemoval()
+                .rescheduleAfter(progress.getRequeueAfterDuration());
+          } else {
+            break;
```

Comment:
Do we need to add a log for requeue?

Reply:
For this one, let me take a look and handle seperately. For now, the corner cases like failures and requeues are not fully tested because it's already handled by StatefulSet mostly.